monai/metrics/metric.py

```diff
@@ -15,121 +15,165 @@
+        The return value is usually a "batch_first" tensor, or a list of "batch_first" tensors.
         """
         raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
 
 
-class Cumulative(ABC):
+class Cumulative:
     """
     Utility class for the typical cumulative computation process based on PyTorch Tensors.
-    It cumulates tensors in the buffer, then sync across distributed ranks and aggregate.
+    It provides interfaces to accumulate values in the local buffers, synchronize buffers across distributed nodes,
+    and aggregate the buffered values.
 
-    To speed up computation with multi-processing, PyTorch programs usually split data to distributed ranks
-    by `DistributedSampler` before an epoch, every rank then computes only based on its own data part and
-    `add` to the buffers in its process. Eventually, sync the values of all ranks to compute the final results.
+    In multi-processing, PyTorch programs usually distribute data to multiple nodes. Each node runs with a subset
+    of the data, adds values to its local buffers. Calling `get_buffer` could gather all the results and
+    `aggregate` can further handle the results to generate the final outcomes.
+
+    Users can implement their own `aggregate` method to handle the results,
+    using `get_buffer` to get the buffered contents.
 
     Note: the data list should have the same length every time calling `add()` in a round,
     it will automatically create buffers according to the length of data list.
 
-    Typically, this class is expected to execute the steps referring to below examples::
+    Typically, this class is expected to execute the following steps:
+
+    .. code-block:: python
+
+        from monai.metrics import Cumulative
+
+        c = Cumulative()
+        c.append(1)  # adds a value
+        c.extend([2, 3])  # adds a batch of values
+        c.extend([4, 5, 6])  # adds a batch of values
+        print(c.get_buffer())  # tensor([1, 2, 3, 4, 5, 6])
+        print(len(c))  # 6
+        c.reset()
+        print(len(c))  # 0
 
-        cum = Cumulative()
-        cum.add(x, y)
-        cum.add(a, b)
-        cum.add(c, d)
-        cum.aggregate()
-        result = cum.get_buffer()  # optional
-        cum.reset()
+    The following is an example of maintaining two internal buffers:
+
+    .. code-block:: python
+
+        from monai.metrics import Cumulative
+
+        c = Cumulative()
+        c.append(1, 2)  # adds a value to two buffers respectively
+        c.extend([3, 4], [5, 6])  # adds batches of values
+        print(c.get_buffer())  # [tensor([1, 3, 4]), tensor([2, 5, 6])]
+        print(len(c))
+
+    The following is an example of extending with variable length data:
+
+    .. code-block:: python
+
+        import torch
+        from monai.metrics import Cumulative
+
+        c = Cumulative()
+        c.extend(torch.zeros((8, 2)), torch.zeros((6, 2)))  # adds batches
```

Comment:
Could you please help explain more about the usage of `extend`?
I didn't get why adding this new API against the `append` or the previous `add`.

Thanks.

Reply:
yes, these APIs will provide flexibility in handling the batch dimension.